File: linkedList/142.linked-list-cycle-ii/142.linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        
        if (head == null) return null;

        ListNode slow = head;
        ListNode fast = head;

        while (fast != null) {
            if (fast.next == null) {
                return null;
            } else {
                fast = fast.next.next;
            }

            slow = slow.next;

            if (slow == fast) {
                break;
            }
        }

        if (fast == null) return null;

        /// 循環リストだとわかったので、それぞれの操作を施す
        slow = head;

        while (slow != fast) {
            slow = slow.next;
            fast = fast.next;
        }
        return slow;
    }
}

File: linkedList/142.linked-list-cycle-ii/142.linked-list-cycle-ii.kt
/*
 * @lc app=leetcode id=142 lang=kotlin
 *
 * [142] Linked List Cycle II
 */

// @lc code=start
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun detectCycle(head: ListNode?): ListNode? {
        if (head == null) return null

        var slow = head
        var fast = head

        while (fast != null && fast?.next != null) {
            slow = slow?.next
            fast = fast?.next?.next
            if (slow == fast) {
                break
            }
        }

        if (fast == null || fast?.next == null) return null

        slow = head

        while (slow != fast) {
            slow = slow?.next
            fast = fast?.next
        }

        return slow 
    }
}
// @lc code=end

